src/rust/session/src/lib.rs

```diff
@@ -241,21 +249,44 @@ impl Session {
     pub fn finalize_response(&mut self) {
         let previous = self.pending_bytes;
         let current = self.write_pending();
-        let len = current - previous;
-
-        self.pending_bytes = current;
 
-        // `len` holds the number of bytes for the finalized response, for empty
-        // 'responses' (eg: memcache NOREPLY), we finalize an empty response to
-        // track the latency. This can be recorded immediately, as there is no
-        // queued response in the buffer.
-        if len == 0 {
-            let now = Instant::now();
-            let latency = (now - self.timestamp()).as_nanos() as u64;
-            REQUEST_LATENCY.increment(now, latency, 1);
-        } else {
-            self.pending_responses.push_back(len);
+        match current.cmp(&previous) {
+            Ordering::Greater => {
+                // We've finalized a response that has some pending bytes to
+                // track. If there's room in the tracking struct, we add it so
+                // we can determine latency later.
+                if self.pending_count < self.pending_responses.len() {
+                    let mut idx = self.pending_head + self.pending_count;
+                    if idx >= self.pending_responses.len() {
+                        idx %= self.pending_responses.len();
+                    }
+                    self.pending_responses[idx] = current - previous;
+                    self.pending_count += 1;
+                }
+            }
+            Ordering::Equal => {
+                // We've finalized a response that is zero-length. This is
+                // expected for empty responses such as when handling memcache
+                // requests which specify `NOREPLY`. Since there are no pending
+                // bytes for a zero-length response, we can determine the
+                // latency now.
+                let now = Instant::now();
+                let latency = (now - self.timestamp()).as_nanos() as u64;
+                REQUEST_LATENCY.increment(now, latency, 1);
+            }
+            Ordering::Less => {
```

Comment:
Why wouldn't this lead to `Ordering:Equal` ? If request handling result in a reduction in shrinkage of the write buffer, seems that'd be hinting at buffer corruption and therefore more severe than a warning?